NIFI-6640 - UNION/CHOICE types not handled correctly

tpalfy opened the pull request:
https://issues.apache.org/jira/browse/NIFI-6640

3 important changes:
1. FieldTypeInference had a bug when dealing with multiple datatypes for
 the same field where some (but not all) were in a wider-than-the-other
 relationship.
 Before: Some datatypes could be lost. String was wider than any other.
 After: Consistent behaviour. String is NOT wider than any other.
2. Choosing a datatype for a value from a ChoiceDataType:
 Before it chose the first compatible datatype as the basis of conversion.
 After change it tries to find the most suitable datatype.
3. Conversion of a value of avro union type:
 Before it chose the first compatible datatype as the basis of conversion.
 After change it tries to find the most suitable datatype.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

tpalfy commented on nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/test/java/org/apache/nifi/avro/TestAvroTypeUtil.java:
```diff
@@ -587,4 +588,49 @@ public void testListAndMapConversion() {
     }
+
+    @Test
+    public void testConvertToAvroObjectWhenFloatVSUnion_INT_FLOAT_ThenReturnFloat() {
```
These two tests can be done more elegantly. Also need to add a version where INT is expected (and maybe test some other types).

markap14 commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+                            RecordFieldType dataType2 = dataTypeMapper.apply(type2).getFieldType();
+
+                            // Moving TIMESTAMP at the front (at least it should precede DATE)
+                            if (dataType1 == RecordFieldType.TIMESTAMP) {
```
This seems a bit convoluted. Does it not make sense to instead just move the TIMESTAMP before DATE in the enum, if this is what we want?

tpalfy replied:
Yeah, that would be the best. I wasn't sure if the current ordering to stay as it is was important for other reasons (didn't wan't to risk regression).
But if you think it's fine to change the ordering I can do that.

markap14 commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+                    boolean compatible;
+
+                    try {
+                        compatible = isCompatibleDataType(valueAsString, dataTypeMapper.apply(type));
```
This can be made much simpler by just simply returning the value. I.e., `return isCompatibleDataType(...);` and then `return false` if an Exception is caught.

tpalfy replied:
I really like to follow the "one return only" convention because I find returning a simple local variable at the end of the method actually helps reading/understanding the code in 2 ways:
1. Easy to find _what_ is being returned and during debugging it's a user-friendly experience to just put a break point on that single line if one wants to know what is returned value.
2. One can simply click on the variable that is being returned and have the IDE highlight it's usages across the method, helping a lot understanding the flow of the logic (as most of the time highlighting the variable that is being returned is basically highlighting the main decision making points in the method where the returning variable is being assigned a value).
This is really a personal preference (although I think it is beneficial to most). But if there's a different convention that is usually used I can apply that as well.

tpalfy replied:
Simplified this as well.

markap14 commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+                    try {
+                        compatible = isCompatibleDataType(valueAsString, dataTypeMapper.apply(type));
+                    } catch (Exception e) {
+                        compatible = false;
```
I'm not sure why any Exception would be thrown from this method. If it is, we should at least be logging it, rather than simply swallowing the Exception.

tpalfy replied:
Indeed no exception should be thrown and probably won't.

I added the try-catch for the general idea that during iterating through all the types, if for whatever reason a non-fitting type behaves strangely and throws an exception it shouldn't abruptly stop the selection process (as the correct type is very likely coming later after the bad-behaving one).

I thought that knowing why a given non-fitting type would throw an exception is not too important.
For example it might be possible that trying to see if a regular string or a number is a timestamp or a date, would involve something  - maybe later on, if not now - that would cause some kind of formatting related exception. We probably wouldn't care about it that much as long as we find the proper type.

But I can add a log line here. It shouldn't happen and if it does it might be useful to know.

tpalfy replied:
Added error log.

markap14 commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -440,12 +539,12 @@ public static DataType inferDataType(final Object value, final DataType defaultT
-        if (value instanceof Object[]) {
-            final Object[] array = (Object[]) value;
-
+        if (value.getClass().isArray()) {
```
Neat! I don't think I realized that `Class.isArray()` and `Array.getLength()` / `Array.get()` were a thing.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:
```diff
@@ -877,6 +878,18 @@ private static Object convertToAvroObject(final Object rawValue, final Schema fi
+                subSchema -> AvroTypeUtil.determineDataType(subSchema)
+        );
+        if (mostSuitableType.isPresent()) {
+            Object convertedVAlue = conversion.apply(mostSuitableType.get());
```
This would be made simpler by simply returning `conversion.apply(...)`

tpalfy replied:
Yeah, I guess even my previous reasoning doesn't hold well as this method already has multiple return statements. (Although during debugging it still can be useful to just add a break point and see immediately the returned value, some simplification could be better.)

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/AbstractTestConversion.java:
```diff
@@ -0,0 +1,396 @@
+    }
+
+    protected String getContent(String postfix) {
+        String content = new String(getByteContent(postfix));
```
We can simply return `new String(getByteContent(posfix));`

tpalfy replied:
Yeah, I guess here half my previous reasoning is moot so in such simple cases it's fine to have a single line that returns an expression.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/AbstractTestConversion.java:
```diff
@@ -0,0 +1,396 @@
+    protected byte[] getByteContent(String postfix) {
+        byte[] content;
+        try {
+            content = Files.readAllBytes(Paths.get("src/test/resources/TestConversions/data.int_float_string." + postfix));
```
We can again just return the value returned by the method call.

tpalfy replied:
Simplified this as well.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/AbstractTestConversion.java:
```diff
@@ -0,0 +1,396 @@
+    }
+
+    protected ArrayList<Map<String, Object>> getRecords(byte[] avroData) throws IOException, MalformedRecordException {
+        RecordReader reader = new AvroReaderWithEmbeddedSchema(new ByteArrayInputStream(avroData));
```
Here too.

tpalfy replied:
This is going to be modified by using a try with resources on the reader.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/AbstractTestConversion.java:
```diff
@@ -0,0 +1,396 @@
+
+    protected ArrayList<Map<String, Object>> getRecords(RecordReader reader) throws IOException, MalformedRecordException {
+        try {
+            ArrayList<Map<String, Object>> records = new ArrayList<>();
```
Is there a reason to declare an `ArrayList` instead of `List`?

tpalfy replied:
Yeah, no, not sure how that happened. :)

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/AbstractTestConversion.java:
```diff
@@ -0,0 +1,396 @@
+
+            return records;
+        } finally {
+            reader.close();
```
Is generally bad form IMO to close a `Closeable` provided to you as an argument. It generally should be closed using a `try-with-resource` by whoever created it.

tpalfy replied:
Good point!

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestConversionWithExplicitSchema.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.ArrayList;
+import java.util.Map;
+
+public class TestConversionWithExplicitSchema extends AbstractTestConversion {
```
Any of these unit tests that are testing conversion from A to B is fine, but they should probably be integration tests, not unit tests. So should be named ConversionWithExplicitSchemaIT

tpalfy replied:
Fair point!

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/schema/inference/FieldTypeInference.java:
```diff
@@ -62,36 +62,44 @@ public void addPossibleDataType(final DataType dataType) {
-            singleDataType = dataType;
-            return;
-        }
+        boolean hasWiderNonString = possibleDataTypes.stream()
```
This method is also evaluated against potentially every field in any record whose schema is being inferred, so performance is important. As such, we should avoid use of `Stream`s and instead prefer imperative style programming.

tpalfy replied:
Replaced Stream api usage.

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/schema/inference/FieldTypeInference.java:
```diff
@@ -62,36 +62,44 @@ public void addPossibleDataType(final DataType dataType) {
+                .isPresent();
+
+        if (!hasWiderNonString) {
+            java.util.Iterator<DataType> possibleDataTypeIterator = possibleDataTypes.iterator();
```
Should import `java.util.Iterator` rather than declaring the full package inline.

tpalfy replied:
Hmm, it must have been a glitch in the Matrix. :)

tpalfy commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
-                if (subType.getFieldType() == RecordFieldType.CHOICE) {
-                    return chooseDataType(value, (ChoiceDataType) subType);
-                }
+        Queue<DataType> possibleSubTypes = new LinkedList<>(choiceType.getPossibleSubTypes());
```
This is kind of similar to a Breath First Search. I'm flattening the Choice datatypes iteratively and add the constituents back to the same data structure I'm iterating over, hence the Queue.

szaboferee commented on nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/record/TestDataTypeUtils.java:
```diff
@@ -324,4 +332,199 @@ public void testConvertDataTypeBigint() {
+        runWithNormalAndReverseInput(this::testChooseDataType, input(dataTypes, value), expected);
+    }
+
+    private <E> void runWithNormalAndReverseInput(BiFunction<Input, E, ?> test, Input input, E expected) {
```
It looks like the first parameter is always the same

tpalfy replied:
Yeah, in the end this `BiFunction` became unnecessary.

tpalfy commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
         }
 
-        return null;
+        List<DataType> compatibleSimpleSubTypes = possibleSimpleSubTypes.stream()
```
Fair point, I'll do a performance overhaul on this method.

tpalfy commented on nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:
```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+    public static <T> Optional<T> findMostSuitableType(Object value, List<T> types, Function<T, DataType> dataTypeMapper) {
+        final Optional<T> mostSuitableType;
+
+        Optional<DataType> inferredDataTypeOptional = Optional.ofNullable(inferDataType(value, null))
```
Refactored.

tpalfy commented:
@markap14 Thanks for your insights! I applied most recommended changes and updated the pull request.

Please note that as part of the performance overhaul in DataTypeUtils.chooseDataType I removed the Set altogether.
This saves us the whole dealing with the set for the cost of doing the same work if the same type occurs more than once.
I think it's a fair price as most of the time we won't encounter type duplication at this point and the performance increase is noticeable.

markap14 commented:
Thanks @tpalfy things are looking much better performance-wise. Things look good to me. Will merge.
